File: programmers/etc/level2/혼자_놀기의_달인.py
def solution(cards):
    answer = []
    n = len(cards)
    for i in range(n):
        tmp = []
        while cards[i] not in tmp:
            tmp.append(cards[i])
            i = cards[i]-1
        answer.append(sorted(tmp) if sorted(tmp) not in answer else [])
    answer.sort(reverse=True, key=len)
    return len(answer[0])*len(answer[1])


cards = [8,6,3,7,2,5,1,4]
print(solution(cards)) # 12

File: Java/programmers/etc/level2/혼자놀기의달인.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 혼자놀기의달인 {
    public int solution(int[] cards) {
        int n = cards.length;
        boolean[] visited = new boolean[n];
        List<Integer> answer = new ArrayList<>();
        
        for (int i=0; i<n; i++) {
            int now = i;
            int cnt = 0;
            while (!visited[now]) {
                cnt++;
                visited[now] = true;
                now = cards[now]-1;
            }
            answer.add(cnt);
        }
        answer.sort((o1,o2)->{return o2-o1;});
        return answer.size() == 1 ? 0 : answer.get(0)*answer.get(1);
    }
}
